[SYCL][ESIMD] Add tests on simd select 1D functions

vasilytric opened the pull request:
Provides tests for select r-value, l-value and for 1D simd_view functions.

vasilytric committed:
[SYCL][ESIMD] Add test on simd rvalue select function

yuriykoch commented on SYCL/ESIMD/api/functional/functions/common.hpp:
```diff
@@ -0,0 +1,55 @@
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides common things for simd functions tests.
```
```suggestion
/// This file provides common code for simd functions tests.
```

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-142364f52b9ed0f64bc04d00a839c547bb9b7a86051c1caa085e720e48ca8265R10)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/common.hpp:
```diff
@@ -0,0 +1,55 @@
+                  const std::string &data_type)
+      : m_data_type(data_type), m_retrieved_val(retrieved_val),
+        m_expected_val(expected_val), m_index(index),
+        m_number_selected_values(number_selected_values), m_stride(stride) {}
```
Would you mind to use a new log approach from #838 with `ESIMD_TESTS_DISABLE_DEPRECATED_TEST_DESCRIPTION_FOR_LOGS` defined?

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-142364f52b9ed0f64bc04d00a839c547bb9b7a86051c1caa085e720e48ca8265L19-R36)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_rvalue.hpp:
```diff
@@ -0,0 +1,204 @@
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides functions for tests on simd rvalue select function.
```
```suggestion
/// This file provides functions for tests on simd select function used as r-value in expression.
```

vasilytric replied:
The new comment can be found [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-ee52ba82babf88cc257b75cc71c61eca3df7aa19a2b6ce7b5ed19349642a56f6R11)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_rvalue.hpp:
```diff
@@ -0,0 +1,204 @@
+
+// Aliases to provide size or stride values to test.
+template <int N> using stride = std::integral_constant<int, N>;
+template <int N> using size = std::integral_constant<int, N>;
```
There are 3 interconnected concerns regarding this code:

1. Please mention the type in these type names explicitly. For example, `size` is too likely to be read as the variable name. 
See https://github.com/intel/llvm-test-suite/blob/intel/SYCL/ESIMD/api/simd_mask.cpp#L27

3. BTW, why not to use a `integer_pack` from `type_coverage.hpp` with `for_all_combinations`?

4. Also please note that `using` doesn't provide a new type, so there is no additional type safety in different aliases. So you might pass a `size<3>` instead of `stride<3>` with no errors.

vasilytric replied:
The `_type` suffix has been added [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-ee52ba82babf88cc257b75cc71c61eca3df7aa19a2b6ce7b5ed19349642a56f6R38-R41)

> BTW, why not to use a integer_pack from type_coverage.hpp with for_all_combinations?

Now we have an self-describe variables like `stride_type`, and by using the `integer_pack` we will iterate over combinations of different sizes and strides, but this iteration is useless 

> Also please note that using doesn't provide a new type, so there is no additional type safety in different aliases. So you might pass a size<3> instead of stride<3> with no errors.

No we don't have a good solution to solve this problem

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_rvalue.hpp:
```diff
@@ -0,0 +1,204 @@
+    size_t alignment_value = alignof(DataT);
+
+    constexpr size_t value_for_increase_ref_data_for_fill = 50;
+    static_assert(std::numeric_limits<char>::max() >
```
Did you mean `signed` or `unsigned` char? The sign of `char` type is implementation-defined.

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-ee52ba82babf88cc257b75cc71c61eca3df7aa19a2b6ce7b5ed19349642a56f6R88)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_rvalue.hpp:
```diff
@@ -0,0 +1,204 @@
+    size_t alignment_value = alignof(DataT);
+
+    constexpr size_t value_for_increase_ref_data_for_fill = 50;
+    static_assert(std::numeric_limits<char>::max() >
```
Looks like we need additional `#include` for `numeric_limits`

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-ee52ba82babf88cc257b75cc71c61eca3df7aa19a2b6ce7b5ed19349642a56f6R22-R23)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_rvalue.hpp:
```diff
@@ -0,0 +1,204 @@
+  constexpr int coefficient_of_division = 3;
+  constexpr int small_offset_value = 1;
+  const int large_offset_value =
+      desired_simd_large_size - std::ceil(2 * desired_simd_large_size / 3);
```
Cannot we replace `ceil` with arithmetic operations and have `constexpr` here as well?

vasilytric replied:
the constexpr function has been added [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-ee52ba82babf88cc257b75cc71c61eca3df7aa19a2b6ce7b5ed19349642a56f6R29-R36) and it has been used [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-ee52ba82babf88cc257b75cc71c61eca3df7aa19a2b6ce7b5ed19349642a56f6R190-R191)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_rvalue.hpp:
```diff
@@ -0,0 +1,204 @@
+      run_test, select_rval,
+      size<desired_simd_large_size / coefficient_of_division>, stride<3>,
+      do_not_use_offset>(all_types, great_size, queue, 0);
+  passed &= for_all_combinations<
+      run_test, select_rval, size<3>,
+      stride<desired_simd_large_size / coefficient_of_division>,
+      do_not_use_offset>(all_types, great_size, queue, 0);
```
What if we have something like this? Would it still work?

```suggestion
  passed &= for_all_combinations<
      run_test, select_rval, size<3>,
      stride<desired_simd_large_size / coefficient_of_division>,
      do_not_use_offset>(all_types, great_size, queue, 42);
```

Could we possibly avoid pairs of template parameter `use_offset`/`do_not_use_offset` and runtime parameter `offset` by using constexpr `offset` values?

vasilytric replied:
The `static_assert` has been added for this case in this commit: 179a8f24c9550018b59b5f10350c5b6a0c6b6faf

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_rvalue.hpp:
```diff
@@ -0,0 +1,204 @@
+};
+
+template <tested_types TestedTypesT>
+bool run_test_with_chosen_data_types(sycl::queue &queue) {
```
Just as a suggestion:
```suggestion
bool run_test_for_types(sycl::queue &queue) {
```

vasilytric replied:
The function has been renamed [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-ee52ba82babf88cc257b75cc71c61eca3df7aa19a2b6ce7b5ed19349642a56f6R183)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_rvalue.hpp:
```diff
@@ -0,0 +1,204 @@
+  }
+};
+
+template <tested_types TestedTypesT>
```
Please, avoid using `T` suffix if it's not a type.

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-ee52ba82babf88cc257b75cc71c61eca3df7aa19a2b6ce7b5ed19349642a56f6R182)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_rvalue.hpp:
```diff
@@ -0,0 +1,204 @@
+    });
+    queue.wait_and_throw();
+
+    // Verify that selected values has been changed to values from referece data
+    // for fill.
+    for (size_t i = offset; i < Stride * NumSelectedElems + offset;
+         i += Stride) {
+      DataT expected = ref_data_for_fill[(i - offset) / Stride];
+      DataT retrieved = result[i];
+      if (expected != retrieved) {
+        passed = fail_test(i, expected, retrieved, data_type);
+      }
+    }
+    // Verify that values, that do not was selected has initial values.
+    for (size_t i = 0; i < NumElems; ++i) {
+      if (i % Stride != 0) {
```
You've actually had presented a great approach with `O(N)` complexity for a different PR. Why not to use it here as well?

vasilytric replied:
The new algorithm has been implemented [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-ee52ba82babf88cc257b75cc71c61eca3df7aa19a2b6ce7b5ed19349642a56f6R134-R155)

vasilytric committed:
[SYCL][ESIMD] Add assert that verify that select will be correct

vasilytric committed:
[SYCL][ESIMD] Structure updates

- Add constexpr ceil function
- Rename function
- Use different algorithm to verify obtained result

vasilytric committed:
[SYCL][ESIMD] Add tests on lvalue select and simd view select

vasilytric committed:
Merge branch 'intel' into select_rval

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+// for std::numeric_limits
+#include <limits>
+
+namespace esimd = sycl::ext::intel::experimental::esimd;
```
```suggestion
namespace esimd = sycl::ext::intel::esimd;
```
Let's move it to the `functions/common.hpp` for this PR. What would you say?

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/875/commits/ee611b346fdc32de475536463be8cb3d45e3f5ea

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+
+constexpr int ceil(float arg) {
+  return (arg > static_cast<int>(arg)) ? arg + 1 : arg;
+}
```
UB here, because there is no check for `arg` out of `int` values range. Not sure we really do need `float` to calculate `details::ceil(2 * desired_simd_large_size / 3);`

Maybe something like this could help?
The following is true for integral `A` and `B`: `details::ceil(A / B) == (A / B) + ((A % B) > 0)`

vasilytric replied:
The `ceil` has been changed in this commit: https://github.com/intel/llvm-test-suite/pull/875/commits/15df4d483843798e6e821ab67f21598727fbc2f4

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+} // namespace details
+
+// Aliases to provide size or stride values to test.
+// This is the syntax sugar and they have the same type.
```
As a suggestion:
```suggestion
// This is the syntax sugar, so no additional type safety here to expect
```
or 
```suggestion
// This is the syntax sugar just for code readability
```

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/03d94c35b4a95e3ce26c99e1197bc50ccdc652aa#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R36)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+  static std::string get_description() { return "select rvalue"; }
+
+  template <typename DataT, int NumElems, int NumSelectedElems, int Stride>
+  static bool call_operator(const DataT *const ref_1, const DataT *const ref_2,
```
Looks like we would be unable to get out the difference between `ref_1` and `ref_2` without reading the entire function body.

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/03d94c35b4a95e3ce26c99e1197bc50ccdc652aa#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R49-R50)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+};
+
+// Descriptor class for the case of calling simd<T,N>::select function.
+struct select_simd_view {
```
```suggestion
struct select_simd_view_rval {
```
Given the naming above. What about l-value?

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/03d94c35b4a95e3ce26c99e1197bc50ccdc652aa#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R92)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+    dst_simd_obj.copy_from(ref_2);
+
+    src_simd_obj.template select<NumSelectedElems, Stride>(offset) =
+        dst_simd_obj;
```
Actually, it's confusing to see `source = destination` assignment

vasilytric replied:
The new name `simd_for_change_values` has been changed [here](https://github.com/intel/llvm-test-suite/pull/875/commits/03d94c35b4a95e3ce26c99e1197bc50ccdc652aa#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R80-R81)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+public:
+  bool operator()(sycl::queue &queue, const std::string &data_type) {
+    static_assert(NumElems >= NumSelectedElems * Stride + Offset &&
+                  "Offset should be less than number selected elements.");
```
Message differs from the expression

vasilytric replied:
Message has been changed [here](https://github.com/intel/llvm-test-suite/pull/875/commits/03d94c35b4a95e3ce26c99e1197bc50ccdc652aa#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R130)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+    static_assert(std::numeric_limits<signed char>::max() >
+                      value_for_increase_ref_data_for_fill + NumElems,
+                  "Value that used for increase ref data for fill plus simd "
+                  "size should be less than char max value.");
```
```suggestion
    constexpr size_t value_for_increase_ref_data_for_fill = 50;
    static_assert(std::numeric_limits<signed char>::max() >
                      value_for_increase_ref_data_for_fill + NumElems);
```
No need to duplicate information. Looks like it's easy to forget update it in two places.

vasilytric replied:
this comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/03d94c35b4a95e3ce26c99e1197bc50ccdc652aa#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3L130-R136)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+    for (size_t i = 0; i < NumElems; ++i) {
+      ref_data_for_fill[i] =
+          initial_ref_data[i] + value_for_increase_ref_data_for_fill;
+    }
```
looks like two calls to the `std::iota` would be enough

vasilytric replied:
the comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/03d94c35b4a95e3ce26c99e1197bc50ccdc652aa#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R145)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+    });
+    queue.wait_and_throw();
+
+    std::vector<size_t> selected_indexsess;
```
```suggestion
    std::vector<size_t> selected_indexes;
```
or
```suggestion
    std::vector<size_t> selected_indices;
```

vasilytric replied:
the comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/03d94c35b4a95e3ce26c99e1197bc50ccdc652aa#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R168)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+    // Push the largest value to avoid the following error: can't dereference
+    // out of range vector iterator.
+    selected_indexsess.push_back(std::numeric_limits<size_t>::max());
+    auto selected_indexsess_ptr = selected_indexsess.begin();
```
```suggestion
    auto next_selected_index = selected_indexes.begin();
```
It's iterator after all

vasilytric replied:
the comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/03d94c35b4a95e3ce26c99e1197bc50ccdc652aa#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R178)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+      // hasn't been selected and changed.
+      if (i < *selected_indexsess_ptr) {
+        DataT expected = initial_ref_data[i];
+        DataT retrieved = result[i];
```
```suggestion
        const DataT& expected = initial_ref_data[i];
        const DataT& retrieved = result[i];
```
Here and below

vasilytric replied:
the comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/03d94c35b4a95e3ce26c99e1197bc50ccdc652aa#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R185-R186)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+    log::fail(TestDescription<NumElems, TestCaseT>(data_type),
+              "Unexpected value at index ", i, ", retrieved: ", retrieved,
+              ", expected: ", expected, ", with size: ", NumSelectedElems,
+              ", stride: ", Stride, ", offset: ", Offset);
```
Looks like size, stride and offset should go to the test description.

vasilytric replied:
The `TestDescription` has been updated in this commit: https://github.com/intel/llvm-test-suite/pull/875/commits/62ef8cdc98df2c8ba5b02911491adc9a7f432d42

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+      log::print_line("Test failed due to type of the object that returns " +
+                      TestCaseT::get_description() +
+                      " operator is not equal to the expected one for simd<" +
+                      data_type + ", " + std::to_string(NumElems) + ">.");
```
```suggestion
      log::fail(TestDescriptionT(data_type), "Unexpected return type");
```
Please note that `TestDescriptionT` should have all test case information logged inside.
So, for example,

- if you have `log::trace<TestDescriptionT>(data_type, other arguments);`
- in `run_test::operator()`
- and define `ESIMD_TESTS_VERBOSE_LOG`
- there would be no equal lines

vasilytric replied:
The `TestDescription` has been updated in this commit: https://github.com/intel/llvm-test-suite/pull/875/commits/62ef8cdc98df2c8ba5b02911491adc9a7f432d42

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+  passed &=
+      for_all_combinations<run_test, SelectT, size_type<1>, stride_type<1>,
+                           offset_type<zero_offset_value>>(types, small_size,
+                                                           queue);
```
Equal calls. Misprint?

vasilytric replied:
yes, duplicate has been deleted

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+  passed &= for_all_combinations<
+      run_test, SelectT,
+      size_type<desired_simd_large_size / coefficient_of_division>,
+      stride_type<3>, offset_type<zero_offset_value>>(types, great_size, queue);
```
Given that `constexpr int coefficient_of_division = 3;` these calls are equal.
What's the meaning behind `coefficient_of_division`?

vasilytric replied:
yes, duplicate has been deleted

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,281 @@
+      stride_type<desired_simd_large_size / coefficient_of_division>,
+      offset_type<zero_offset_value>>(types, great_size, queue);
+
+  passed &= for_all_combinations<
+      run_test, SelectT,
+      size_type<desired_simd_large_size - small_offset_value>,
+      stride_type<desired_simd_small_size>, offset_type<small_offset_value>>(
+      types, great_size, queue);
+  passed &=
+      for_all_combinations<run_test, SelectT,
+                           size_type<desired_simd_large_size / 3>,
+                           stride_type<2>, offset_type<large_offset_value>>(
+          types, great_size, queue);
```
Actually, I would suggest at least to use something like `run_with_size_stride_offset` if it makes sense.. Maybe I'm missing something, but it's definitely hard to read and need to be refactored in any way.

vasilytric replied:
The `run_with_size_stride_offset` has been implemented in this commit: https://github.com/intel/llvm-test-suite/pull/875/commits/4fe9429145e7b85bbf4457dc385e8d749ba08d74

But I am not sure that this is a good solution

vasilytric committed:
[SYCL][ESIMD] Update ceil function

Now it works with two int args to avoid UB float value may be greater
than int values range

vasilytric committed:
[SYCL][ESIMD] Move namespace alias into functions/common.hpp

vasilytric committed:
[SYCL][ESIMD] Small structure changes

- Update comments
- Rename variables
- Rename files

vasilytric committed:
Merge branch 'intel' into select_rval

vasilytric committed:
[SYCL][ESIMD] Change TestDescription and context of using it

vasilytric committed:
[SYCL][ESIMD] Add wrapper-function to run the test

vasilytric committed:
[SYCL][ESIMD] Disable coverage for char types

vasilytric committed:
[SYCL][ESIMD] Apply clang-format

vasilytric committed:
[SYCL][ESIMD] Update description for select_simd_view_rval

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,292 @@
+#include "../shared_element.hpp"
+#include "common.hpp"
+
+// for std::ceil
+#include <cmath>
```
Just to make sure - do we still need it?

vasilytric replied:
no, removed in this commit: https://github.com/intel/llvm-test-suite/pull/875/commits/7f9db3eebf5ec8ff52475f0920fad0fb0875349f

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,292 @@
+
+namespace details {
+
+constexpr int ceil(int a, int b) { return (a % b) > 0 ? a / b + 1 : a / b; }
```
Not really required, but just as suggestion:
```suggestion
constexpr int ceil(int a, int b) { return ((a % b) > 0) ? (a / b + 1) : (a / b); }
```

vasilytric replied:
this comment applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/6e3c354f9cdd7463f3b5749cc615b9ff59eaa9ca#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R29-R31)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_1d_select.hpp:
```diff
@@ -0,0 +1,292 @@
+
+template <typename SelectT, int NumSelectedElems, int Stride, int Offset,
+          typename... T, int... Values>
+bool run_with_size_stride_offset(
+    sycl::queue &queue, const named_type_pack<T...> &types,
+    const unnamed_type_pack<std::integral_constant<int, Values>...> &sizes) {
```
Why not just `std::forward` parameters passed? We are not using `queue`, `types`, `sizes` or `Values` here anyway.
```suggestion
bool run_with_size_stride_offset(ArgsT &&... args) {
```

vasilytric replied:
this comment applied [here](https://github.com/intel/llvm-test-suite/pull/875/commits/6e3c354f9cdd7463f3b5749cc615b9ff59eaa9ca#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R222-R223)

vasilytric committed:
[SYCL][ESIMD] Remove unused library

vasilytric committed:
[SYCL][ESIMD] Use parameter pack in run_with_size_stride_offset

This lets do not mention all arguments explicitly

vasilytric committed:
[SYCL][ESIMD] Use namespace to access to esimd API

vasilytric commented:
@sndmitriev Would you mind to review this PR?

vasilytric commented:
@kbobrovs Would you mind to merge this PR?
